Improve forced commands for RP2040 and custom vid/pid

will-v-pi opened the pull request:
Allow forced commands with RP2040 on Windows - they will not deactivate the mass storage device, to ensure Zadig still works

Also fix forced commands with custom vid/pid (as they will change in bootsel mode), by removing vid/pid filters after reboot and just filtering on serial number

Fixes #260 and #261

will-v-pi committed:
Remove vid/pid filter when using forced commands

The vid/pid will typically change when switching to BOOTSEL mode, so remove those filters and just filter on serial number if present

Fixes #261

will-v-pi committed:
Improve handling of RP2040 devices after reboot

Calling GET_INFO when not available sometimes causes issues with forced commands

Also improve libusb error messages when debugging

will-v-pi committed:
Add rp2040 workaround to allow forced reboots, and improve behaviour with no vid/pid filtering

will-v-pi committed:
Readme update for --rp2040 option

will-v-pi committed:
Add Zadig message to failed RP2040 forced reboots

Also encourage -f/F on Windows, rather than a separate reboot

lurch commented on main.cpp:
```diff
@@ -8762,6 +8772,13 @@ int main(int argc, char **argv) {
                         settings.bus = -1;
+                        if (settings.pid != -1 || settings.vid != -1) {
+                            // also skip vid/pid filtering, as that should change in BOOTSEL mode, and could be white-labelled on RP2350
+                            settings.pid = -1;
```
Seems slightly confusing that the comment-line talks about "vid/pid filtering", but the code-line only mentions `settings.pid`?

lurch commented on main.cpp:
```diff
@@ -8762,6 +8772,13 @@ int main(int argc, char **argv) {
                         settings.address = -1;
                         settings.bus = -1;
+                        if (settings.pid != -1 || settings.vid != -1) {
+                            // also skip vid/pid filtering, as that should change in BOOTSEL mode, and could be white-labelled on RP2350
+                            settings.pid = -1;
+                            // still filter for rpi vid/pid if we don't have a serial number, as that is an RP2040 running a no_flash binary, so will
+                            // have a standard rpi vid/pid in BOOTSEL mode
+                            settings.vid = settings.ser.empty() ? -1 : 0;   // 0 means skip vid/pid filtering entirely, -1 means filter for rpi vid/pid
```
Perhaps something like this is slightly less confusing? :man_shrugging: 
```suggestion
                            if (settings.ser.empty()) {
                                // still filter for rpi vid/pid if we don't have a serial number, as that is an RP2040 running a no_flash binary, so will
                                // have a standard rpi vid/pid in BOOTSEL mode
                                settings.vid = -1; // -1 means filter for rpi vid/pid
                                settings.pid = -1;
                            } else {
                                // also skip vid/pid filtering, as that should change in BOOTSEL mode, and could be white-labelled on RP2350
                                settings.vid = 0; // 0 means skip vid/pid filtering entirely
                                settings.pid = -1;
                            }
```

will-v-pi replied:
Have resolved this with something similar to this

lurch commented:
I _think_ the "Use RP2040 workarounds on Windows when using custom vid/pid (ignored on other platforms)" stuff is assuming that if you're running a UF2 on an RP2350 with a custom VID/PID, then you'll have **also** modified the RP2350's OTP to report the same VID/PID in BOOTSEL mode? But I guess that _might_ not always be the case? :thinking:

will-v-pi commented:
> I _think_ the "Use RP2040 workarounds on Windows when using custom vid/pid (ignored on other platforms)" stuff is assuming that if you're running a UF2 on an RP2350 with a custom VID/PID, then you'll have **also** modified the RP2350's OTP to report the same VID/PID in BOOTSEL mode? But I guess that _might_ not always be the case? 🤔

Nope, the workaround is just to not disable the MSD, nothing to do with the actual vid/pid - if you have RP2040 stdio_usb vid/pid it autodetects that and does it automatically, but if you're using custom vid/pid then you need to tell it you have an RP2040 as it can't detect it from the vid/pid

lurch commented:
> > I _think_ the "Use RP2040 workarounds on Windows when using custom vid/pid (ignored on other platforms)" stuff is assuming that if you're running a UF2 on an RP2350 with a custom VID/PID, then you'll have **also** modified the RP2350's OTP to report the same VID/PID in BOOTSEL mode? But I guess that _might_ not always be the case? 🤔
> 
> Nope, the workaround is just to not disable the MSD, nothing to do with the actual vid/pid - if you have RP2040 stdio_usb vid/pid it autodetects that and does it automatically, but if you're using custom vid/pid then you need to tell it you have an RP2040 as it can't detect it from the vid/pid

Okay, and how is the situation different on RP2350?

will-v-pi commented:
> Okay, and how is the situation different on RP2350?

RP2350 bootrom has the correct MS_OS_20 USB descriptors so works on Windows out of the box - whereas RP2040 doesn’t so requires drivers installed by Zadig, which only work properly when the RP2040 in bootsel mode has the MSD enabled

lurch commented:
> RP2350 bootrom has the correct MS_OS_20 USB descriptors so works on Windows out of the box - whereas RP2040 doesn’t so requires drivers installed by Zadig, which only work properly when the RP2040 in bootsel mode has the MSD enabled

Ahhhh! I knew it'd probably be something obvious that I'd forgotten :joy:  Thanks for the explanation.

lurch commented on main.cpp:
```diff
@@ -593,6 +594,7 @@ auto device_selection =
         (option("--vid") & integer("vid").set(settings.vid).if_missing([] { return "missing vid"; })) % "Filter by vendor id" +
         (option("--pid") & integer("pid").set(settings.pid)) % "Filter by product id" +
         (option("--ser") & value("ser").set(settings.ser)) % "Filter by serial number"
+        + option("--rp2040").set(settings.force_rp2040) % "Use RP2040 workarounds on Windows when using custom vid/pid (ignored on other platforms)"
```
```suggestion
        + option("--rp2040").set(settings.force_rp2040) % "Use RP2040-specific workarounds when using a custom vid/pid on Windows (ignored on other platforms)"
```
? (it niggled me slightly that "on other platforms" was so far away from "Windows", that it wasn't entirely clear what "other platforms" was referring to :wink: )

will-v-pi committed:
Tidy up comments around vid/pid filter disable

will-v-pi committed:
Tweak help text

Co-authored-by: Andrew Scheller <andrew.scheller@raspberrypi.com>

will-v-pi committed:
Update readme with help text

will-v-pi committed:
Add Zadig note to the readme

This is cleaner than directing to a document name that must then be searched for the word Zadig

lurch commented on README.md:
```diff
@@ -1630,3 +1658,8 @@ this requirement (see the [hello_usb](https://github.com/raspberrypi/pico-exampl
 If you ctrl+c out of the middle of a long operation, then libusb seems to get a bit confused, which means we aren't able
 to unlock our lockout of USB MSD writes (we have turned them off so the user doesn't step on their own toes). Simply running
 `picotool info` again will unlock it properly the next time (or you can reboot the device).
+
+### Zadig
+To communicate with RP2040 in BOOTSEL mode on Windows, you will need to install a driver. To do this, download and run [Zadig](http://zadig.akeo.ie), select `RP2 Boot (Interface 1)` from the dropdown box and select `WinUSB` as the driver, and click on the "Install Driver" button. Wait for the installation to complete - this may take a few minutes.
```
```suggestion
### Zadig

To communicate with RP2040 in BOOTSEL mode on Windows, you will need to install a driver. To do this, download and run [Zadig](http://zadig.akeo.ie), select `RP2 Boot (Interface 1)` from the dropdown box and select `WinUSB` as the driver, and click on the "Install Driver" button. Wait for the installation to complete - this may take a few minutes.
```
(minor niggle)

will-v-pi committed:
Update README.md

Co-authored-by: Andrew Scheller <andrew.scheller@raspberrypi.com>

kilograham commented on main.cpp:
```diff
@@ -8703,6 +8705,13 @@ int main(int argc, char **argv) {
                     }
+
+                    if (settings.vid == 0 && !settings.ser.empty() && !devices[dr_vidpid_bootrom_ok].empty()) {
+                        // Searching with no vid/pid filtering (ie opening all devices) can cause issues, so stop
```
i have no idea what this means - maybe just improve the comment a  bit

will-v-pi replied:
I've added more detail to the comment

kilograham commented on main.cpp:
```diff
@@ -8797,13 +8806,14 @@ int main(int argc, char **argv) {
-                                    fail(ERROR_NOT_POSSIBLE,
-                                        "Forced commands do not work with RP2040 on Windows - you can force reboot into BOOTSEL mode via 'picotool reboot -f -u' instead.");
+                                if (desc.idProduct == PRODUCT_ID_RP2040_STDIO_USB || settings.force_rp2040) {
+                                    disable_mask = 0;   // enable MSC interface so Zadig works correctly
```
again - not much context for why this is the case - again maybe improve comment

will-v-pi replied:
Here’s a more detailed explanation
```suggestion
                                    // the Zadig driver should be setup for the device in BOOTSEL mode with no interfaces disabled,
                                    // as there are none disabled when you plug it in while holding the BOOTSEL button
                                    disable_mask = 0;
```

lurch replied:
Maybe "there are none disabled" -> "all the interfaces are enabled"?

will-v-pi replied:
```suggestion
                                    // the Zadig driver should be setup for the device in BOOTSEL mode with no interfaces disabled,
                                    // as all the interfaces are enabled when you plug it in while holding the BOOTSEL button
                                    disable_mask = 0;
```

lurch replied:
Minor tweak, but maybe "while holding the BOOTSEL button" -> "while holding down the BOOTSEL button" ?

kilograham commented on main.cpp:
```diff
@@ -8841,6 +8851,18 @@ int main(int argc, char **argv) {
                         settings.address = -1;
                         settings.bus = -1;
+                        if (settings.pid != -1 || settings.vid != -1) {
+                            // vid/pid filtering was enabled, but should change in BOOTSEL mode, so needs to be disabled
```
should -> may ?

will-v-pi replied:
```suggestion
                            // vid/pid filtering was enabled, but may change in BOOTSEL mode, so needs to be disabled
```

kilograham commented on main.cpp:
```diff
@@ -610,6 +611,7 @@ auto device_selection =
         (option("--vid") & integer("vid").set(settings.vid).if_missing([] { return "missing vid"; })) % "Filter by vendor id" +
         (option("--pid") & integer("pid").set(settings.pid)) % "Filter by product id" +
         (option("--ser") & value("ser").set(settings.ser)) % "Filter by serial number"
+        + option("--rp2040").set(settings.force_rp2040) % "Use RP2040-specific workarounds when using a custom vid/pid on Windows (ignored on other platforms)"
```
im still not quite sure I'd know what this does - basically this is a "you must pass this when using a custom VID/PID on windows and want to do a force reboot"?

kilograham replied:
actually force_rp2040 is a confusing name for a variable anyway, and "--rp2040" should probably be the name if and only if it means "please assume the device is an rp2040"

will-v-pi replied:
I think "please assume the device might be an rp2040" is what it means - for SDK stdio_usb it can tell from the PID so you don’t need to pass it, but if you have a custom VID/PID you need to pass `--rp2040` to enable the workarounds.

If you pass `--rp2040` when using an RP2350 on Windows, it will still work fine, but it will keep the mass storage interface enabled in bootsel mode.

Maybe this is more explicit:
```suggestion
        + option("--rp2040").set(settings.force_rp2040) % "Assume the device might be an RP2040 - this is only required when using a custom vid/pid with an RP2040 on Windows, to enable the RP2040 Windows workarounds, and is ignored on other operating systems"
```

kilograham replied:
yeah but i dont think the user knows what the workarounds are, so its basically "this is required when using RP2040 with a custom vid/pid on Windows to workaround a Windows problem and is ignored on other platforms" (IDK if workaround a windows problem is true)

lurch replied:
Also, "Assume the device is an RP2040" might be better wording than "Assume the device might be an RP2040" ?

Following on from Graham's suggestion, maybe "to enable the RP2040 Windows workarounds" could be removed entirely, or perhaps replaced with "to enable RP2040-specific workarounds" ?

will-v-pi replied:
How about this - I think "operating systems" is clearer than "platforms", as it doesn't depend on the pico_platform
```suggestion
        + option("--rp2040").set(settings.force_rp2040) % "Assume the device is an RP2040 - this is only required when using a custom vid/pid with an RP2040 on Windows, and is ignored on other operating systems"
```

kilograham reviewed (approved):
I'm approving pending some comment/help text clarification - then go ahead and merge

will-v-pi committed:
Improve comments and description

will-v-pi committed:
Update readme

will-v-pi committed:
Merge remote-tracking branch 'origin/develop' into force-no-vid-pid

will-v-pi committed:
fix readme

will-v-pi committed:
review fixup - add "down"
